### ISSUE TITLE
Наложение горизонталей

### ISSUE THREAD

**Issue by wladich** [wladich/nakarte#403]
Хорошо бы добавить слой наложения горизонталей. 
Для этого желательно найти готовый слой и чтобы владельцы были не сильно против его использования на стороннем ресурсе.
Делать свой статический не получится, т.к. на 18 масштабе он будет слишком много места занимать.
Можно пордумать про динамический.

**Comment by wladich** [comment]
Результаты изысканий. 

OpenTopoMap использует горизонтали, которые можно попросить у OpenSnowMap. Т.е. документация у них в этой части не соответствует действительности. Процесс получения горизонталей и OpenSnowMap не задокументирован (или я не нашел), но, скорее всего, для широт S56-N60 используется NASA SRTM V3 1 arc second, для остальных - неизвестно, но точно не ViewFinderPanoramas.

Примечательные баги на ОТМ:
1. https://nakarte.me/#m=14/47.44864/13.07776&l=Otm (кратеры)
2. https://nakarte.me/#m=16/47.51641/12.98895&l=Otm (бугры на озере)
3. https://nakarte.me/#m=16/46.77517/12.80718&l=Otm (кратер)
4. https://nakarte.me/#m=14/65.58462/-14.01886&l=Otm (кратер севернее N60)
5. https://nakarte.me/#m=15/47.48606/13.61416&l=Otm (кратер)

* viewfinderpanoramas 3 arc - багов нет. Но разрешение на севере низкое
* NASA SRTM V3 1 arcsecond - Заклеены все дырки. Все баги присутствуют (кроме N4, т.к. он севернее N60)
* NGA  SRTM  1 Arc-Second Global (версия непонятная, т.к. в описании не указана, в сторонней статье называлась 2.1, в ссылках на тайлы указана v3) - на месте багов обширные области no-data, сильно больше по размеру.
* ALOS - 1 arc-second, полное покрытие земли, дырки заклеены NASA SRTM 1 arc, ASTER GDEM и др. - много бугров или ям высотой 30-50 м, причем и в равнинных областях. Багов 1-5 нет
* ASTER GDEM - arc-second, полное покрытие земли, тоже есть бугры, но меньше, зато есть баги N2, 5, т.е., похоже, что NASA SRTM заклеивалось частично этими данными

Горизонтали в Google Terrain больше всего похожи на ViewFinderPanoramas, в т.ч. в районе бага N4 - как по детальности, так и по смещению.

**Comment by wladich** [comment]
Вывод - надо использовать ViewFinderPanoramas 3 arc-second. 
Можно попробовать сделать статические векторные тайлы до 13 или 14 уровня, возможно, что размер будет не запредельный.

**Comment by slazav** [comment]
Да, я очень рекомердую ViewFinderPanoramas. Какая-нибудь Якутия там сделана по советским двухкилометровкам, и не слишком точна. Но для практических целей достаточно. Существенных ошибок я там вообще не видел.

**Comment by wladich** [comment]
С генерацией более-менее красивых горизонталей разобрался, получается база Postgis порядка 600-700 Мб. За счет прореживания раза в 2, наверное, можно уменьшить, но на сервер такое тащить тяжеловато.

Смотрел на mapbox vector tiles: mapnik умеет их делать, но я не понял зачем. Т.е. чем тайлы, нарезанные mapnik-ом отличаются от нарезанных любым другим инструментом прямо из векторных данных, т.к. никаких элементов оформления я в них не увидел. Конкретно мне нужно получить расставленные подписи горизонталей.

Т.е. при использовании векторных тайлов надписи предлагается расставлять в браузере, чего плагины для leaflet, кажется, делать не умеют. Библиотека Mapbox GL JS как-то умеет, но, наверное, хуже mapnik-а и выковыривать из неё нужные куски и делать свой рендеринг настроения нет.

**Comment by wladich** [comment]
Пример, если наложить на стандартный OSM (горизонтали через 20 м)
![Screenshot at 2020-11-30 21-48-12](https://user-images.githubusercontent.com/1458202/100663220-44d5d300-3356-11eb-94b1-e7ac767fd413.png)

Можно сделать горизонтали через 100 м, тогда размер базы должен уменьшиться в 10 раз (с прореживанием):
![Screenshot at 2020-11-30 21-49-56](https://user-images.githubusercontent.com/1458202/100663237-4901f080-3356-11eb-9bd1-2d142517278c.png)

**Comment by wladich** [comment]
A вот полезность с космоснимками уже не такая очевидная:
![Screenshot at 2020-11-30 21-55-04](https://user-images.githubusercontent.com/1458202/100663565-c75e9280-3356-11eb-9168-fc9952deb3c7.png)


Что заставляет задуматься, насколько вообще этот слой будет востребован - кроме OSM, google и яндекс все карты имеют свой рельеф.

**Comment by slazav** [comment]
Вот, кстати, как у меня это выглядит:
![srtm1](https://user-images.githubusercontent.com/37889/100666820-8b74fe80-3351-11eb-8709-ffe2f22d3b51.jpg)

Но я таким никогда не пользуюсь, мне важнее цветные картинки (ты их видел, конечно). Та что с высотами менее полезная, хотя иногда приятно посмотреть относительную высоту вершин и горных районов. А с уклонами я очень люблю для планировании маршрута рассматривать:

![srtm2](https://user-images.githubusercontent.com/37889/100667229-21108e00-3352-11eb-9269-d404b1efe41d.jpg)

![srtm3](https://user-images.githubusercontent.com/37889/100667243-25d54200-3352-11eb-9199-25f610cd9d85.jpg)

**Comment by wladich** [comment]
Для коллекции: другой вариант стиля со снимком
![Uploading Screenshot at 2020-12-01 17-44-54.png…]()

**Comment by slazav** [comment]
Белая линия - интересная тема! По картинке кажется, что она у тебя только с одной стороны (что странно, конечно). Может, надо просто рисовать широкую белую и поверх коричневую поуже?

**Comment by wladich** [comment]
Нет, она специально с одной стороны. Потому как стоит задача:
1. Сделать максимально заметной на светлом и темном фоне.
2. Минимально скрывать детали нижележащего слоя.

С п.1 односторонняя и двусторонняя обводка справляются одинаково, а для п.2 односторонняя выигрывает.

Кстати, аналогичный приём использован в километровой сетке при печати.
